### DISCUSSION TITLE
parse query from string and execute with params

### DISCUSSION THREAD

**Discussion by bluebrown** [sqlalchemy/sqlalchemy#7404] [Usage Questions] [answered]
Hi, is it possible to compile a query and parse its string represantation again somewhere else? I want to send query and params to celery and assemble them again.

```python
q = (
    select([EmailTemplate])
    .where(EmailTemplate.name.in_(["a.pdf", "b.pdf"]))
    .where(EmailTemplate.sender == "foo")
    .compile(dialect=postgresql.dialect())
)
# I want to send those to celery, so they must be string and dict
params = q.params
str_q = str(q)

# celery should use the query string and execute it with the params
engine = create_engine(str(POSTGRES_URL))
with Session(engine, future=True) as session:
    session.execute(str_q, params)
```

It's strange because it puts double % signs in the final query.

```python
[SQL: SELECT ...
FROM emailtemplate 
WHERE emailtemplate.name IN (__[POSTCOMPILE_name_1]) AND emailtemplate.sender = %%(sender_1)s]
```

The result is a syntax error.

```python
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near "%"
LINE 3: ..._[POSTCOMPILE_name_1]) AND emailtemplate.sender = %(sender_1...
```

**Comment by zzzeek** [comment]
there is a query serializer extension at https://docs.sqlalchemy.org/en/14/core/serializer.html?highlight=serializer#module-sqlalchemy.ext.serializer which might solve this problem more simply for you.

the caveat with this extension is that as it relies on Python pickle, more complex queries may fail to serialize correctly.  you'd get an error message immediately if this were the case, however, so if a particular query serializes fine then it's safe to use.

otherwise, your issue is two things.  first off, IN expressions produce a special form known as [post compile](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#all-in-expressions-render-parameters-for-each-value-in-the-list-on-the-fly-e-g-expanding-parameters) which if rendered has to be converted to individual parameters before being passed to the dialect.   This can be performed by adding the [render_postcompile](https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#rendering-postcompile-parameters-as-bound-parameters) flag to your string compilation. 

next issue, psycopg2's parameter style is known as [pyformat](https://www.python.org/dev/peps/pep-0249/#paramstyle).  The parameter style used by text() is to use [colons](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=text#sqlalchemy.sql.expression.text).   you need to convert all parameters that look like `%(name)s` into colon style, e.g. `:name`.  Otherwise, all the `%` characters you have in there are escaped before being passed to the psycopg2 dialect as they are [assumed to be literal % characters](https://www.psycopg.org/docs/usage.html#passing-parameters-to-sql-queries).

**Reply by bluebrown** [reply]
thank you, thats some useful information. I will experiment with that.

**Reply by bluebrown** [reply]
It's serializing fine, but I am getting errors while deserializing it.

```python
engine = create_engine(url)

q = (
    select([WarrantyRequest.uuid])
    .where(WarrantyRequest.status.in_(["OPEN", "OK"]))
    .where(WarrantyRequest.installation_date.between("1984-05-23", "2018-12-18"))
    .where(WarrantyRequest.source == "OL")
    .where(WarrantyRequest.partner_address_additional.is_(None))
)

pickled = dumps(q)
to_execute = loads(pickled, engine=engine)
```

```
Traceback (most recent call last):
  File "/wolf-warranty-backend/test.py", line 30, in <module>
    to_execute = loads(pickled, engine=engine)
  File "/wolf-warranty-backend/.venv/lib/python3.9/site-packages/sqlalchemy/ext/serializer.py", line 177, in loads
    return unpickler.load()
  File "/wolf-warranty-backend/.venv/lib/python3.9/site-packages/sqlalchemy/ext/serializer.py", line 155, in persistent_load
    return metadata.tables[table].c[colname]
AttributeError: 'NoneType' object has no attribute 'tables'
```

When I pass the metadata instead, I get a different error,

```python
to_execute = loads(pickled, MetaData(bind=engine))
```

```
Traceback (most recent call last):
  File "/wolf-warranty-backend/test.py", line 30, in <module>
    to_execute = loads(pickled, MetaData(bind=engine))
  File "/wolf-warranty-backend/.venv/lib/python3.9/site-packages/sqlalchemy/ext/serializer.py", line 177, in loads
    return unpickler.load()
  File "/wolf-warranty-backend/.venv/lib/python3.9/site-packages/sqlalchemy/ext/serializer.py", line 155, in persistent_load
    return metadata.tables[table].c[colname]
KeyError: 'warrantyrequest'
```

**Reply by zzzeek** [reply]
> It's serializing fine, but I am getting errors while deserializing it.
> 
> ```python
> engine = create_engine(url)
> 
> q = (
>     select([WarrantyRequest.uuid])
>     .where(WarrantyRequest.status.in_(["OPEN", "OK"]))
>     .where(WarrantyRequest.installation_date.between("1984-05-23", "2018-12-18"))
>     .where(WarrantyRequest.source == "OL")
>     .where(WarrantyRequest.partner_address_additional.is_(None))
> )
> 
> pickled = dumps(q)
> to_execute = loads(pickled, engine=engine)

you don't need to pass "engine" to loads(), you arent using bound metadata and this is a deprecated feature.  you do need to pass the MetaData() object that has your tables in it.

> When I pass the metadata instead, I get a different error,
> 
> ```python
> to_execute = loads(pickled, MetaData(bind=engine))

OK that's not *the* `Metaddata`, that's *a* `MetaData`.  You need to, in the Celery process, import your models / tables and use the `MetaData` that has the table definitions in it, otherwise, it can't deserialize because it can't locate the tables that are named in your SQL construct, hence the error you are getting.   The point of the serializer is that it does not send whole model/table definitions into the serialized string.

**Reply by bluebrown** [reply]
Ok thanks, I will try to do that, and see if it works.

**Reply by bluebrown** [reply]
OK, that worked too. I have passed the same metadata I would pass to alembic for migrations. This will run a parameterized query, right?

```python
from db.base import Base
conn.execute(loads(pickled, metadata=Base.metadata)).all()
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

> I want to send query and params to celery and assemble them again.

For that use case I would consider whatever it would not be better to pass the input that you use to generate the query to celery and generate the query in the worker there.

It this remains the preferred use case you can do as mike suggested regarding renderer_postcompile, meaning something like:
```py
compiler = q.compile(engine, compile_kwargs={"render_postcompile": True})
query_str = compiler.string
query_parms = compiler.params
go_query.send(query_str, query_params)  # send to celery

# in celery
@celery.task
def go_query(query_str, query_params):
  # TODO: validation!?!
  with engine.connect() as conn:
    conn.exec_driver_sql(quert_str, query_params)
```
(the celery bit are purely pseudo code, don't remember how to use it from memory)

**Reply by bluebrown** [reply]
Hi, thanks for the suggestion. I will try that out with the compiled query. Regarding generating the query in the worker, that is what I am doing now. But it's not as easy as it seems. For example, tuples are not preserved, I guess because they are not valid JSON and therefore converted to lists. 

I was using this function, but now I have issued with the tuple. 
```python
def where_from_params(model: Any, query: ClauseElement, params: Optional[Dict[str, Any]] = None) -> ClauseElement:
    q = query
    if params:
        for key, val in params.items():
            if val is None:
                continue
            if isinstance(val, tuple):
                q = q.where(getattr(model, key).between(*val))
            elif isinstance(val, list):
                q = q.where(getattr(model, key).in_(val))
            elif val == "NULL":
                q = q.where(getattr(model, key).is_(None))
            else:
                q = q.where(getattr(model, key) == val)
    return q
```

 I am getting those params from a large collection of query parameters in an HTTP handler, and I am validating them with pydantic. That's why, for example, a date range might end up in a tuple. 

I am also not sure how to create an or clause with this approach, either, although I actually don't need an OR at the moment.

But you may be right, maybe I should refactor this to parse the query string only in the worker and validate there with models or something similar.

**Reply by bluebrown** [reply]
Do you mean when I compile the query I lose the safety of parameterized queries? Because I want to keep it. That why I though sending the query and the params separately so that alchemy will send them both to the Postgres server to let it do the validation part.

**Reply by bluebrown** [reply]
I have tried the suggested way, but I guess the problem is still the Postgres dialect. If I keep it I have %s in the query and I get an error saying dict is not a sequence. 

```
WHERE warrantyrequest.installation_date BETWEEN %s AND %s
```

```
cursor.execute(statement, parameters)
TypeError: dict is not a sequence
```


If I dont pass the engine, so compile doesn't know the dialect, the query looks actually fine and is matching the params, 
but it's throwing another error.

```
WHERE warrantyrequest.installation_date BETWEEN :installation_date_1 AND :installation_date_2]
[parameters: {'installation_date_1': '1980-01-01T00:00:00', 'installation_date_2': '2020-01-01T00:00:00'}]
```
```
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near ":"
LINE 3: WHERE warrantyrequest.installation_date BETWEEN :installatio...
```

**Reply by zzzeek** [reply]
OK, the ":colon" format is what's used by [text()](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=text#sqlalchemy.sql.expression.text).  so you have a SQL string with colons in it, that's good.  now put that string inside of text() and execute that.

```
     result = connection.execute(text(my_string_with_colons))
```

**Reply by bluebrown** [reply]
OK, wrapping it in text did the trick.  I am right when thinking that this has still the safety benefit of a parameterized query?

```python
q = (
    select([WarrantyRequest.uuid])
    .where(WarrantyRequest.status.in_(["OPEN", "OK"]))
    .where(WarrantyRequest.installation_date.between("1984-05-23", "2018-12-18"))
    .where(WarrantyRequest.source == "AP")
    .where(WarrantyRequest.partner_address_additional.is_(None))
)

compiler = q.compile(compile_kwargs={"render_postcompile": True})
params = compiler.params
query = compiler.string

with engine.connect() as conn:
    res = conn.execute(text(query), **params).all()
```
